content/articles/constructors_in_kotlin/index3.md

```diff
@@ -0,0 +1,205 @@
+           println("Email = ${mail}")
+           //excutes first
+     }
+}    
+     fun main(args:Array<String>){
+     val mystudent = myStudent("Alex",200) 
+    
+}
+```
+The output of the above code is:
+
+```
+//excuted first
+Name = Alex
+Id =200 
+Email = Alex@gmail
+//excuted last
+Name=Alex 
+Id=200
+```
+         
+
+
+
+
+
+### Init blocks
+Init is short for initializer. It is a block of code that is executed every time the class is instantiated. They are used when you need to add a block of code that cannot be added to the constructor, simplifying a lot for the programmer. Init blocks are called immediately after the primary constructor. To declare one, use the `init` keyword:
+
+```kotlin
+init
+{ 
+    //code to be executed
+}
+```
+
+Here is an example of a primary constructor with an init block:
+
+```kotlin
+class person(name:String,id:Int){ 
+    val pName :String 
+    val pId:Int 
+    init{ 
+        pName= name.capitalize() 
+        pId=id 
+        println("Name = ${pName}") 
+        println("Id=${pId}") 
+        } 
+    } fun main(args:Array<String>){ 
+        val Person=person("Alex",200) 
+}
+
+```
+
+When a person object is created, the parameters name and id accept the values `Alex` and `200`. The properties name and id are not properties of the Person class because they are not preceded by `val` or `var`. After the constructor is called, the initializer block is executed which initializes `pname` and `pId`. When run this would be the output:
+
+```
+Name = Alex
+Id =200
+```
+
+### The difference between a primary constructor and a secondary constructor
+The main difference between a primary and secondary constructor is that the former resides in the class declaration header, whereas the latter is declared inside the Kotlin class body. Additionally, a secondary constructor may call the primary constructor.
+
+### Best practices for object-oriented programming
+The clean use of classes is required for effective Kotlin. We learned about constructors above but never discussed best practices with them. We can make the following changes to our code to make it more readable while using constructors:
+- Minimize the visibility of properties, i.e., when using the private, public, protected keywords. It allows you to control where your class members can be accessed from.
+- Combine property declaration. This is achieved by taking a property and setting it to a constructor parameter.
+- Improve variable naming. This is achieved by using variable naming rules; your variable names should: 
+  - Be based on the variable's purpose so it's clear what the variable is for.
+  - Have no spaces between words. Ex. myClass.
+  - Be greater than a single character.
+  - Not use special prefixes or suffixes (e.g. `name_`, `mName`, `s_name`, and `kName`), except in the case of backing properties.
+  
+
+Note: The init block is called just after the primary constructor, and the secondary constructor is called using the `this` keyword.
```

Comment:
Why is this under *best practices for object-oriented programming*? It doesn't belong here move it to be under a more fitting subheader.